cataclysmbnteam/Cataclysm-BN
# [Port] Add in-game progression diary

#### Summary



#### Purpose of change
Fixes #1917 
An in-game way to take notes was something missing in the game, so when I saw that someone already made it in DDA (+extra features), I just had to port it



#### Describe the solution
Port the diary
https://github.com/CleverRaven/Cataclysm-DDA/pull/52372 (Original PR)
https://github.com/CleverRaven/Cataclysm-DDA/pull/53576 (Fix typos)
https://github.com/CleverRaven/Cataclysm-DDA/pull/55045 (INFO menu option)
https://github.com/CleverRaven/Cataclysm-DDA/pull/53582 (Better translation support for head text)
https://github.com/CleverRaven/Cataclysm-DDA/pull/54720 (Clean code, fix typos)
https://github.com/CleverRaven/Cataclysm-DDA/pull/56253 (Add more robust code that fixes several bugs)
https://github.com/CleverRaven/Cataclysm-DDA/pull/57321 (Change the pop op after an edit to Yes by default)

**A few personal modifications**
* 1 - Fix the head text bugging through the second diary page on small to medium screens
* 2 - Fix the Info display on small screens (phones) by allowing up to 3 lines to be displayed
- 3 - Complete rework of UI, removing the padding to improve lisibility and maximizing the used space (all screens)
_Comparison between the current BN and DDA versions_
![image](https://user-images.githubusercontent.com/71428793/195409462-34f17148-086f-4880-af36-dc9cdec37d90.png)
- 4 - Add a maximum width limit and centering the UI after reaching the limit (medium to big screens)
- 5 - Replace the time since last entry by a shorter, universal (no translation needed for hours/minutes/seconds) and more precise version

_At the limit_
![image](https://user-images.githubusercontent.com/71428793/195412650-5f414541-90ae-4bcb-8d37-ed4a30d28a48.png)

_Above the limit_
![image](https://user-images.githubusercontent.com/71428793/195412380-09c3a593-4868-4c6c-acb6-ea1be57066fa.png)
- 6 - Fix many typos and viariable names that didn't use snake case
- 7 - Add martial arts progression tracking (what martial arts are known at the start, then the new ones learnt)
- 8 - Change export format from .txt to .md. .md is for [markdown](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/quickstart-for-writing-on-github) and it's basically readable by everything that can read .txt. With this I could add a way to better seperate the entries in the exported file
- 9 - Add maximum bionic power progression tracking



#### Testing
I did the same tests as in the original PR:
- create, edit an entry
- create, delete a page
- export the diray
- save and load
- resize the game window
- die and write last entry
- verify that the progression is tracked
(Since last changes)
- try to go outside the window and find a way to crash the game
- write special chars and save/load game, export diary
